Remove REDISMODULE_ prefixes and introduce compatibility header

PingXie opened the pull request:
Fix #146 

Removed REDISMODULE_ prefixes from the core source code to align with the new SERVERMODULE_ naming convention. Added a new 'redismodule.h' header file to ensure full backward compatibility with existing modules. This compatibility layer maps all legacy REDISMODULE_ prefixed identifiers to their new SERVERMODULE_ equivalents, allowing existing Redis modules to function without modification.

PingXie commented:
One high level question.

Should we use SERVERMODULE_* or VALKEYMODULE_*? I have been using the generic "server" term everywhere in this PR, including both file names and APIs. Now on a second thought, I feel that "VALKEYMODULE_" would probably work better. I am leaving the PR as-is for now but curious to hear the thoughts from others.

natoscott commented:
+1 for VALKEYMODULE_ prefix.

zuiderkwast commented:
Yeah, this header file is usually copied to the module's repo. Using a generic name like "servermodule" is not appropriate in this context, i.e. outside of valkey.

When documenting this, we should somehow explain that the valkeymodule API is a superset of redismodule API as it was in redis 7.2. I believe it can be confusing to module authors, especially if they want to be compatible with various forks at the same time.

madolson commented:
> +1 for VALKEYMODULE_ prefix.

I also agree with this. I think server is fine when we're just talking about agnostic stuff, but this is the ValkeyModule system now.

madolson commented on src/servermodule.h:
```diff
@@ -0,0 +1,1695 @@
+#define SERVERMODULE_CTX_FLAGS_MULTI (1<<1)
+/* The instance is a master */
+#define SERVERMODULE_CTX_FLAGS_MASTER (1<<2)
+/* The instance is a slave */
+#define SERVERMODULE_CTX_FLAGS_SLAVE (1<<3)
+/* The instance is read-only (usually meaning it's a slave as well) */
```
My one maybe bold take, we should update this:
```suggestion
/* The instance is a replica */
#define SERVERMODULE_CTX_FLAGS_REPLICA (1<<3)
/* The instance is read-only (usually meaning it's a replica as well) */
```

We said we don't use slave except for compatibility, and this isn't strictly compatibility.

zuiderkwast replied:
Change MASTER flags too then? Or shall we leave that for later/never?

zuiderkwast replied:
If we do change to REPLICA for this flag, then the docs for migrating from redismodule to valkeymodule needs to mention this explicitly, apart from the s/REDISMODULE/VALKEYMODULE/g and s/RedisModule/ValkeyModule/g.

@PingXie Can you add this "migration guide" as a comment near the top of valkeymodule.h?

madolson replied:
I thought about MASTER, but we didn't pick a new term. If we can align on that new term I would like to change it to.

PingXie replied:
>I thought about MASTER, but we didn't pick a new term. If we can align on that new term I would like to change it to.

I picked "primary" as it seems to have been established in both the code base and the documentation. 

>@PingXie Can you add this "migration guide" as a comment near the top of valkeymodule.h?

Let me add this in the next commit

PingXie replied:
done

madolson commented on src/servermodule.h:
```diff
@@ -0,0 +1,1695 @@
+#define SERVERMODULE_NODE_ID_LEN 40
+#define SERVERMODULE_NODE_MYSELF     (1<<0)
+#define SERVERMODULE_NODE_MASTER     (1<<1)
+#define SERVERMODULE_NODE_SLAVE      (1<<2)
```
```suggestion
#define SERVERMODULE_NODE_REPLICA      (1<<2)
```

PingXie commented:
SGTM. Will align to "valkeymodule"

madolson commented on src/valkeymodule.h:
```diff
@@ -0,0 +1,1711 @@
+    return VALKEYMODULE_OK;
+}
+
+#define RedisModule_Assert(_e) ((_e)?(void)0 : (RedisModule__Assert(#_e,__FILE__,__LINE__),exit(1)))
```
```suggestion
#define ValkeyModule_Assert(_e) ((_e)?(void)0 : (RedisModule__Assert(#_e,__FILE__,__LINE__),exit(1)))
```

Not sure if you meant to cover this define or meant to hit it later.

PingXie replied:
yeah - I was planning to do just one PR for everything but it became big and then I changed my mind and was planning to stage the changes in multiple PRs but I guess that also creates this confusion about the final goal. I guess I will keep patching up this PR then.

madolson commented on src/valkeymodule.h:
```diff
@@ -0,0 +1,1711 @@
+/*
+ * valkeymodule.h
+ *
+ * This header file is renamed from valkeymodule.h to reflect the new naming conventions adopted in Valkey.
```
Seems like an outdated comment.

PingXie replied:
>Seems like an outdated comment.

lol. I added it on purpose to explain why this file exists in the first place and how it is supposed to be used. I have tried hard but I just can't seem to imagine what context a first time valkey module writer who happens to be a veteran redis module developer would need when they see this file. Maybe I should move this comment into the new redismodule.h instead?

madolson replied:
I think it's fine, I think it should just have been "This header file is renamed from redismodule.h". However, I think just omitting this comment and we can document it later makes sense.

madolson reviewed (commented):
The defines mostly look correct to me. I'm a bit concerned that this is touching a lot of unnecessary files we'll need to hit on the backport (mainly the module .c files).

PingXie commented:
> he defines mostly look correct to me. I'm a bit concerned that this is touching a lot of unnecessary files we'll need to hit on the backport (mainly the module .c files).

Agreed and this is still work in progress. @zuiderkwast any concern with moving this out of our 7.2.4 release and into the next one? I will continue patching up the remaining "redis" references.

PingXie commented:
All changes are in. With this PR, we should be both ABI and source compatible with existing Redis modules. New modules can be written using Valkey only interfaces.

`test-sanitizer-address` is complaining about memory leaks. Will take a look tomorrow.

zuiderkwast commented:
> The defines mostly look correct to me. I'm a bit concerned that this is touching a lot of unnecessary files we'll need to hit on the backport (mainly the module .c files).

If it gets hard to backport this change, then as a backup plan we can include it in our next major version instead. It's not that much of a deal. We'll still have "Redis" in log entries and error replies in our 7.2.4 version.

zuiderkwast commented:
> @zuiderkwast any concern with moving this out of our 7.2.4 release and into the next one?

No concern with that. It's more safe to postpone it actually.

madolson commented on src/module.c:
```diff
@@ -33,8 +33,8 @@
- * Each function starting with RM_ and preceded by a block comment is included
- * in the API documentation. To hide an RM_ function, put a blank line between
+ * Each function starting with VK_ and preceded by a block comment is included
+ * in the API documentation. To hide an VK_ function, put a blank line between
```
```suggestion
 * in the API documentation. To hide a VK_ function, put a blank line between
```

Isn't english great.

zuiderkwast replied:
Not VM for ValkeyModule?

PingXie replied:
You are right. VM is closer to RM. Will update

madolson commented on src/module.c:
```diff
@@ -33,8 +33,8 @@
  * The comments in this file are used to generate the API documentation on the
```
This file is far too large to realistically review right now. I'm going to assume everything in this file was just a find/replacement outside of the few strategic compatibility hooks.

zuiderkwast replied:
@madolson I'm going through it. I've made it half-ways so far. I'm not being very strict about correct English. It's code comments, not end-user docs.

PingXie replied:
load and unload hooks are updated. Others are indeed just find/replace. Maybe I should run clang-format on module.c now? :-)

madolson replied:
Eh, we're completely destroying git history, we can do it later and destroy it again.

madolson commented on src/redismodule.h:
```diff
@@ -1,1695 +1,719 @@
 
-#endif /* REDISMODULE_CORE */
-#endif /* REDISMODULE_H */
+#include "valkeymodule.h"
```
Just a note about when we write compatibility. This will not always work, if someone was just copying the file and actually using it for codegen or something. AFAIK the redismodule-rs does this on the .h file.

zuiderkwast replied:
Not a big deal IMO. Users can copy it from redis 7.2.4 instead.

PingXie replied:
I am also less concerned about this. the (build) errors are self-explanatory IMO. I also left sufficient comments in redismodule.h to explain its purpose. Lastly, I would expect module developers to use the redismodule.h copy from the redis tree to build redis modules.

madolson reviewed (approved):
The implementation makes theoretical sense to me for module compatibility. The only thing I found missing was the unload module hook, I would be okay to fast follow that in a separate PR, there are also no existing modules that use it outside of the TLS compatibility one.

zuiderkwast commented:
We have to think about redis version vs valkey version. The API contains REDISMODULE_COMMAND_INFO_VERSION, which is used in some APIs. We may want to track both versions somehow.

madolson commented:
> We may want to track both versions somehow.... or do we?

I honestly think we should introduce a "good" API tracking system for modules, unlike what is used today. I think a lot of modules use the version number to determine which APIs are present which I don't really like.

PingXie commented:
> I honestly think we should introduce a "good" API tracking system for modules, unlike what is used today. I think a lot of modules use the version number to determine which APIs are present which I don't really like.

Agreed. I don't think we should use API version at all. These module APIs are no different than the commands like "HGET" etc. They should be evolved independently with full backward compatibility.

PingXie commented:
> The implementation makes theoretical sense to me for module compatibility. The only thing I found missing was the unload module hook, I would be okay to fast follow that in a separate PR, there are also no existing modules that use it outside of the TLS compatibility one.

I will take a quick look into the unload. I have also narrowed the memory leak issue to a minimum repro.

zuiderkwast commented:
> > I honestly think we should introduce a "good" API tracking system for modules, unlike what is used today. I think a lot of modules use the version number to determine which APIs are present which I don't really like.
> 
> Agreed. I don't think we should use API version at all. These module APIs are no different than the commands like "HGET" etc. They should be evolved independently with full backward compatibility.

Yeah, but even if we do introduce a better method, some modules will still use `RedisModule_GetServerVersion()`. Do we want it to return the redis version (7.2.4) or the valkey version? If we choose the latter, it can be confusing for modules what a number like 8.2.0 means.... So we can freeze it to 7.2.4 maybe?

madolson commented:
> So we can freeze it to 7.2.4 maybe?

I think we can decide this later. For now we should and will return 7.2.4.

PingXie commented:
> > The implementation makes theoretical sense to me for module compatibility. The only thing I found missing was the unload module hook, I would be okay to fast follow that in a separate PR, there are also no existing modules that use it outside of the TLS compatibility one.
> 
> I will take a quick look into the unload. I have also narrowed the memory leak issue to a minimum repro.

turns out that the unload hook is the reason of the leaks :-).

I think this PR is almost ready. Addressing the remaining comments...

madolson reviewed (approved):
Just reviewed the diff, and LGTM

zuiderkwast commented on src/valkeymodule.h:
```diff
@@ -0,0 +1,1698 @@
+ * This header file is forked from redismodule.h to reflect the new naming conventions adopted in Valkey.
+ * 
+ * Key Changes:
+ * - Symbolic names have been changed from VALKEYMODULE_* to VALKEYMODULE_* to align with the
```
From what to what? Fixme

zuiderkwast commented on src/valkeymodule.h:
```diff
@@ -0,0 +1,1698 @@
+VALKEYMODULE_API const char *(*ValkeyModule_GetCurrentCommandName)(ValkeyModuleCtx *ctx) VALKEYMODULE_ATTR;
+VALKEYMODULE_API int (*ValkeyModule_RegisterDefragFunc)(ValkeyModuleCtx *ctx, ValkeyModuleDefragFunc func) VALKEYMODULE_ATTR;
+VALKEYMODULE_API void *(*ValkeyModule_DefragAlloc)(ValkeyModuleDefragCtx *ctx, void *ptr) VALKEYMODULE_ATTR;
+VALKEYMODULE_API ValkeyModuleString *(*ValkeyModule_DefragModuleString)(ValkeyModuleDefragCtx *ctx, ValkeyModuleString *str) VALKEYMODULE_ATTR;
```
DefragValkeyModuleString please. That's what it defrags.
